src/EventStore.Core/Services/Storage/ReaderIndex/IndexBackend.cs

```diff
@@ -24,18 +29,69 @@ public interface IIndexBackend<TStreamId> : IIndexBackend {
+		// used only when cache is configured in terms of "max items" rather than "max bytes".
+		public const long StreamInfoCacheUnitSize = 1000;
+
+		private const int LastEventNumberCacheRatio = 60;
+		private const int StreamMetadataCacheRatio = 100 - LastEventNumberCacheRatio;
+
+		private const string LastEventNumberCacheName = "LastEventNumber";
+		private const string StreamMetadataCacheName = "StreamMetadata";
+
 		private SystemSettings _systemSettings;
 
-		public IndexBackend(ObjectPool<ITransactionFileReader> readers,
-			int lastEventNumberCacheCapacity,
-			int metadataCacheCapacity) {
-			Ensure.NotNull(readers, "readers");
+		public IndexBackend(ObjectPool<ITransactionFileReader> readers, ISizer<TStreamId> streamIdSizer, ICacheSettings streamInfoCacheSettings) {
+			Ensure.NotNull(readers, nameof(readers));
+			Ensure.NotNull(streamIdSizer, nameof(streamIdSizer));
 
 			_readers = readers;
-			_streamLastEventNumberCache = new LRUCache<TStreamId, EventNumberCached>(lastEventNumberCacheCapacity);
-			_streamMetadataCache = new LRUCache<TStreamId, MetadataCached>(metadataCacheCapacity);
+			_streamIdSizer = streamIdSizer;
+
+			var lastEventNumberCacheCapacity =
+				streamInfoCacheSettings.InitialMaxMemAllocation * LastEventNumberCacheRatio / 100;
+			var streamMetadataCacheCapacity =
+				streamInfoCacheSettings.InitialMaxMemAllocation * StreamMetadataCacheRatio / 100;
+
+			_streamLastEventNumberCache = new DynamicLRUCache<TStreamId, EventNumberCached>(lastEventNumberCacheCapacity, CalcLastEventNumberItemSize);
+			_streamMetadataCache = new DynamicLRUCache<TStreamId, MetadataCached>(streamMetadataCacheCapacity, CalcMetadataItemSize);
+
+			Log.Information("{name} cache capacity set to ~{numEntries:N0} bytes.", LastEventNumberCacheName, lastEventNumberCacheCapacity);
+			Log.Information("{name} cache capacity set to ~{numEntries:N0} bytes.", StreamMetadataCacheName, streamMetadataCacheCapacity);
+
+			streamInfoCacheSettings.GetMemoryUsage = () => _streamLastEventNumberCache.Size + _streamMetadataCache.Size;
+			streamInfoCacheSettings.UpdateMaxMemoryAllocation = UpdateMaxMemoryAllocation;
+		}
+
+		private int CalcLastEventNumberItemSize(TStreamId streamId, EventNumberCached eventNumberCached) =>
+			_streamIdSizer.GetSizeInBytes(streamId) + EventNumberCached.Size + EventNumberCached.DictionaryEntryOverhead;
+
+		private int CalcMetadataItemSize(TStreamId streamId, MetadataCached metadataCached) =>
+			_streamIdSizer.GetSizeInBytes(streamId) + metadataCached.Size + MetadataCached.DictionaryEntryOverhead;
+
+		private void UpdateMaxMemoryAllocation(long allocatedMem) {
+			UpdateCapacity(_streamLastEventNumberCache, LastEventNumberCacheName,
+				allocatedMem * LastEventNumberCacheRatio / 100);
+			UpdateCapacity(_streamMetadataCache, StreamMetadataCacheName,
+				allocatedMem * StreamMetadataCacheRatio / 100);
+		}
+
+		private delegate void LogAction(string msg, params object[] args);
```

Comment:
can use `Log.Write` and pass in the level rather than using a delegate to switch between two methods

Reply:
fixed in: e06f87cac61d79fde696581e1a760d0cf14fbb23